sklearn/tree/_tree.pyx

```diff
@@ -1931,6 +1931,17 @@ cdef _build_pruned_tree(
             is_leaf = leaves_in_subtree[orig_node_id]
             node = &orig_tree.nodes[orig_node_id]
 
+            # protect against an infinite loop as a runtime error, when leaves_in_subtree
+            # are improperly set where a node is not marked as a leaf, but is a node
+            # in the original tree. Thus, it violates the assumption that the node
+            # is a leaf in the pruned tree, or has a descendant that will be pruned.
+            if (not is_leaf and node.left_child == _TREE_LEAF
+                    and node.right_child == _TREE_LEAF):
+                raise ValueError(
```

Comment:
I'm surprised this does not need the gil to raise this error.

Also, is there any performance impact with doing this check?

Reply:
I can run a before/after experiment, but my initial suspicion would be no since it's a simple if/else. 

However, if it does, I wonder how we can introduce a check before actually carrying out the pruning because it seems like a reasonable mistake to make even in sklearn future development, and it ends up crashing ones computer cuz nodes are just added repeatedly.

Reply:
I think the `if` condition should be okay. I'm more concerned about adding a `raise ValueError` and Cython doing something strange.

The safer alternative is to set `rc=-2` and break. Then raise a different error here:

https://github.com/scikit-learn/scikit-learn/blob/5600faab452b0adb1f4090219012aa3ae6208c66/sklearn/tree/_tree.pyx#L1961-L1962